app/src/main/java/org/p2p/wallet/common/EncryptedSharedPreferences.kt

```diff
@@ -0,0 +1,255 @@
+package org.p2p.wallet.common
+
+import androidx.core.content.edit
+import android.content.SharedPreferences
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import timber.log.Timber
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty
+import org.p2p.core.crypto.Hex
+import org.p2p.wallet.common.crypto.keystore.DecodeCipher
+import org.p2p.wallet.common.crypto.keystore.EncodeCipher
+import org.p2p.wallet.common.crypto.keystore.KeyStoreWrapper
+
+class EncryptedSharedPreferences(
+    private val keyStoreWrapper: KeyStoreWrapper,
+    private val sharedPreferences: SharedPreferences,
+    private val gson: Gson
+) {
+    var loggingTag: String = "EncryptedSharedPreferences"
+
+    fun saveString(key: String, data: String) {
+        tryWithLog(key) {
+            val encodedData = keyStoreWrapper.encode(key, data)
+            sharedPreferences.edit { putString(key, encodedData) }
+        }
+    }
+
+    fun saveString(key: String, data: String, cipher: EncodeCipher) {
+        tryWithLog(key) {
```

Comment:
tryWithLog it's not obvious
you can inject logger to the func, and name it as saveStringAndLog, to avoid using callbacks

Reply:
sure, let me think of something here